src/util/cilfacade.ml

```diff
@@ -79,6 +79,81 @@ let end_basic_blocks f =
   let thisVisitor = new allBBVisitor in
   visitCilFileSameGlobals thisVisitor f
 
+class loopUnrollingVisitor = object
+  inherit nopCilVisitor
+  method! vstmt s =
+    (* All the duplicated stmts need to be re-created or the sid will fail*)
+    let newsid st = { st with sid = st.sid } in
```

Comment:
This looks fishy. I guess the point is to create a copy of `st`, which is physically unequal to it? Is there some guarantee that the OCaml compiler won't just optimize this to identity?

Also, if this is used to copy the body of the loop however many times necessary, the copies need to have distinct `sid`s because CFG nodes are compared based on the `sid` of the statement. If they have the same `sid`s then the CFG won't contain any copies. Or am I wrong?

Reply:
>  I guess the point is to create a copy of st, which is physically unequal to it?

Yes. It is an idiom that is used in several places in CIL. I didn't find a more explicit way to make a copy of a record with mutable members, but according not the manual this indeed creates a (shallow) copy, so I think the compiler may not optimize this out.

Reply:
> If they have the same sids then the CFG won't contain any copies. Or am I wrong?

According to the documentation http://goblint.in.tum.de/assets/goblint-cil/api/Cil.html#TYPEstmt this should only be computed later, right?

> A number (>= 0) that is unique in a function. Filled in only after the CFG is computed.

Reply:
You're right, the OCaml manual says this creates a shallow copy, so it should be fine. But I think there should be a comment to clarify what it's achieving, because otherwise someone might be tempted at some point to "optimize" it away.

Indeed, the statements are really numbered by `computeCFGInfo`. So considering my remark below about doing this after `prepareCFG` to normalize breaks, continues, etc. This visitor should be invoked between the two.